modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -516,13 +522,32 @@ private CompletableFuture<Void> processTxRecoveryAction(TxRecoveryMessage reques
         boolean transactionAlreadyFinished = txMeta != null && isFinalState(txMeta.txState());
 
         if (transactionAlreadyFinished) {
-            return nullCompletedFuture();
+            if (txMeta.locksReleased()) {
+                // Locks released - nothing to do.
+                return nullCompletedFuture();
+            }
+            // Locks were not released - send cleanup.
+            if (txMeta.enlistedPartitions().isEmpty()) {
```

Comment:
by the way, if you call `txManager.finish` like below, you will never get empty list here: even in case of recovery you will send a tx finish request to commit partition with one partition in `enlistedPartitions` collection: the commit partition itself.